Mathlib/Analysis/InnerProductSpace/Adjoint.lean

```diff
@@ -308,18 +308,36 @@ theorem conj_starProjection {T : E →L[𝕜] E} (hT : IsSelfAdjoint T)
 
 end IsSelfAdjoint
 
+namespace ContinuousLinearMap
+
+variable {T : E →L[𝕜] E} [CompleteSpace E]
+
 open ContinuousLinearMap in
-theorem IsIdempotentElem.hasOrthogonalProjection_range [CompleteSpace E]
-    {p : E →L[𝕜] E} (hp : IsIdempotentElem p) : (LinearMap.range p).HasOrthogonalProjection :=
-  have := hp.isClosed_range.completeSpace_coe
-  .ofCompleteSpace _
+theorem IsIdempotentElem.adjoint_range_eq_range_of (hT : IsIdempotentElem T)
+    (h : (LinearMap.range T)ᗮ = LinearMap.ker T) :
+    LinearMap.range (adjoint T) = LinearMap.range T := by
+  have := hT.hasOrthogonalProjection_range
+  have := hT.star.hasOrthogonalProjection_range
+  rw [← Submodule.orthogonal_orthogonal (LinearMap.range (adjoint T)),
+    orthogonal_range, adjoint_adjoint, ← h, Submodule.orthogonal_orthogonal]
+
+/-- An idempotent operator `T` is self-adjoint iff `(range T)ᗮ = ker T`. -/
+theorem IsIdempotentElem.isSelfAdjoint_iff_orthogonal_range (h : IsIdempotentElem T) :
+    IsSelfAdjoint T ↔ (LinearMap.range T)ᗮ = LinearMap.ker T :=
```

Comment:
ah, can you please add an extensionality lemma for linear idempotents (and continuous ones too)? That is,
```lean
lemma LinearMap.IsIdempotentElem.ext_iff {S T : E →ₗ[𝕜] E} (hS : IsIdempotentElem S)
    (hT : IsIdempotentElem T) :
    S = T ↔ range S = range T ∧ ker S = ker T :=
  sorry

alias ⟨_, LinearMap.IsIdempotentElem.ext⟩ := LinearMap.IsIdempotentElem.ext_iff
```
Once you do that the proof of this lemma should start
```
  rw [isSelfAdjoint_iff, h.ext_iff h.star]
```
and then some shenanigans to turn `range T = range (star T) ∧ ker T = ker (star T)` into `(range T)ᗮ = ker T`.